### ISSUE TITLE
Generate unique model class which extends CRUDService instead of creating at runtime

### ISSUE THREAD

**Issue by craicoverflow** [aerogear/graphback#1059]
A CRUDService is dynamically generated at runtime using the db-specific APIs in `@graphback/runtime-knex` and `@graphback/runtime-mongo`. 

This presents a number of drawbacks:

1. This assumes all models are mapped to a single data-source. In simple scenarios this is likely, but it is likely in large schemas to have more than one data source.
2. If a user wants to customise their CRUD implementation for their models, these runtime APIs cannot be used.

The CRUD abstraction makes it possible for users to customise their own CRUD implementation by creating a custom CRUDService. This works well for an application where the CRUD implementation requirements stays the same across all models. 

However once the customisation requires varies between models it means having to write, configure and maintain multiple services just to add minor customisations.

**NoteCRUDService**

```ts
async function findAll(context: any) {
    const allNotes = await super.findAll(context);

    const notDeletedNotes = allNotes.filter((v: _Note) => !v.deletedAt);

    return notDeletedNotes;
}
```

**CommentCRUDService**

```ts
async function findAll(context: any) {
    if (context.userRole !== "ADMIN") {
        throw new Error("Invalid permissions");
    }

    return await super.findAll(context);
}
```

## Solution

By passing responsibility of creating CRUD services and data providers to a runtime library, we are making it even harder for users to customise their application.

As an alternative we could generate models which extend the CRUDService.

**models/Note.ts**

```ts
import { CRUDService, GraphbackDataProvider, GraphbackPubSub, GraphbackPage } from '@graphback/runtime';
import { GraphQLObjectType } from 'graphql';
import { Note as _Note } from '../generated-types';

export class Note extends CRUDService {
  public constructor(modelType: GraphQLObjectType, db: GraphbackDataProvider, subscriptionConfig: GraphbackPubSub) {
    super(modelType, db, subscriptionConfig)
  }

  public async findAll(context: any): Promise<_Note[]> {
    const allNotes: _Note[] = await super.findAll(context);

    const notDeletedNotes = allNotes.filter((v: _Note) => !v.deletedAt);

    return notDeletedNotes;
  }
}
```

A drawback of this approach is that the services would need to be configured from the server code. Below is a really rough idea of how this could look. This would be a stateless function which would always be regenerated and should not be modified by the user. I don't think it would be possible to have this in a runtime package.

**models/createModels.ts**

```ts
async function createModels(schema: GraphQLSchema, defaultDataProvider: GraphbackDataProvider, dbProviderOverrides: { [modelName: string]: GraphbackDataProvider } = {}, pubSub: PubSub): Promise<{ [serviceName: string]: CRUDService }> {
  const noteModel = schema.getType('Note') as GraphQLObjectType

  ## this would be in a separate file
  const noteConfig = {
    name: "Note",
    pubSub: {
      publishCreate: false,
      publishUpdate: false,
      publishDelete: false,
    },
  }

  // customise dataprovider at model level
  const noteDbProvider = dbProviderOverrides.Note || defaultDataProvider

  const NoteService = new Note(noteModel, noteDbProvider, {
    pubSub,
    ...noteConfig.pubSub
  })

  return {
    Note: NoteService
  }
}
```

**src/index.ts**

```ts
const noteModel = schema.getType('Note') as GraphQLObjectType

const defaultDbProvider = new PgKnexDBDataProvider(noteModel, db);
const customDbProvider = new PgKnexDBDataProvider(noteModel, db);

const models = await createModels(schema, defaultDbProvider, { Note: customDbProvider }, pubSub);
 
const apolloServer = new ApolloServer({
  typeDefs: typeDefs,
  resolvers,
  context: {
    ...context,
    models
  },
  playground: true,
})
```

**Comment by craicoverflow** [comment]
cc @wtrocki this is just an idea and it is open for debate.

**Comment by wtrocki** [comment]
TL;DR approach makes sense interms of codegen but:

+ This approach makes it simple to introduce auth models
+ This approach is way cleaner to write the data

- There is challenge of overriding the files thru code generation
- There is a challenge of generating this content
- This brings us closer to the ORM
- Helpers for runtime still need to be refined

I think we need to refine it to find sweet spot between runtime helper and code generation. I will think about it for the next days.


CC @darahayes

**Comment by darahayes** [comment]
The general idea of having individual models that can then be extended/modified/overridden seems good to me. I think it could be worth prototyping. I do have one question though.

> A drawback of this approach is that the services would need to be configured from the server code. Below is a really rough idea of how this could look. This would be a stateless function which would always be regenerated and should not be modified by the user. I don't think it would be possible to have this in a runtime package.

Could you expand on that a little bit more and explain why it couldn't be done at runtime? Is it because we need to know the names/locations of those models up front and have code that explicitly imports and creates them?

Also another minor thing I noticed. We pass the Model `GraphQLObjectType` down into the DataProvider classes. And the only reason is because of a once off creation of some field name -> database table name mapping as part of the constructor. If we moved towards the approach with individual models, you probably wouldn't need to do this because the CrudService could instead provide that mapping to the dataservice. Either through a setter method or instead of passing the already created DataProvider instance to the CrudService, you could probably pass the constructor for the DataService instead. (either the default one or a custom one), and the mapping could be provided that way.

I think overall this is a really solid idea and it's probably worth prototyping.

**Comment by wtrocki** [comment]
> I think it could be worth prototyping. 

With support for edits?

**Comment by craicoverflow** [comment]
> Could you expand on that a little bit more and explain why it couldn't be done at runtime? Is it because we need to know the names/locations of those models up front and have code that explicitly imports and creates them?

It actually is most definitely possible. I guess I was just unsure how feasible it would be to work with both our code generation and runtime usecases. I can see now that it _should_ work for both.

Here is a realllly rough prototype that works, written fully from the server code (but can be ported to graphback package).

```ts
async function createModels(modelsDir: string, schema: GraphQLSchema, defaultDataProvider: GraphbackDataProvider, dbProviderOverrides: { [modelName: string]: GraphbackDataProvider } = {}, pubSub: PubSub): Promise<{ [serviceName: string]: CRUDService }> {
  const models = require(modelsDir);

  delete models.modelConfigs;

  const services = {};
  for (const modelName of Object.keys(models)) {
    const modelType = schema.getType(modelName) as GraphQLObjectType

    const dbProvider = dbProviderOverrides[modelName] || defaultDataProvider;

    dbProvider.setModel(modelName);

    const modelService: GraphbackCRUDService = new models[modelName](modelType, dbProvider, {
      pubSub,
      ...{
        name: [modelName],
        pubSub: {
          publishCreate: false,
          publishUpdate: false,
          publishDelete: false,
        },
      }
    });

    services[modelName] = modelService;
  }

  return services;
}
```

Given that the `modelsDir` and `schema` can be gotten from the `.graphlrc` config, we could realistically have an API as simple as this:

```ts
import { createRuntime } from 'graphback';

const pubSub = new PubSub();

const db = Knex(...);

const customDbProvider = new PgKnexDBDataProvider(db);

const { typeDefs, resolvers, models } = createRuntime(defaultDbProvider, { Note: customDbProvider }, pubSub);

const apolloServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: context => {
        ...context,
        models: services
    }
})
```

> 
> Also another minor thing I noticed. We pass the Model `GraphQLObjectType` down into the DataProvider classes. And the only reason is because of a once off creation of some field name -> database table name mapping as part of the constructor. If we moved towards the approach with individual models, you probably wouldn't need to do this because the CrudService could instead provide that mapping to the dataservice. Either through a setter method or instead of passing the already created DataProvider instance to the CrudService, you could probably pass the constructor for the DataService instead. (either the default one or a custom one), and the mapping could be provided that way.

Yeah, agreed completely. There is no other reason to have the base type except for the type name and building the mapping, but this does not really need to happen here.

**Comment by darahayes** [comment]
@craicoverflow nice one for the more generic `createModels` idea. I was thinking of something similar. This is a really small thing but if I'm understanding it properly, we'd probably need to fix [this issue first](https://github.com/aerogear/graphback/issues/1028) if we want to be able to read the models in the way that you suggest. I'd be happy to work on that.

**Comment by craicoverflow** [comment]
If we decide to generating individual models files which extend CRUDService I reckon we might need a new ModelServiceGenerator plugin. `models.ts` is generated in the ResolverGeneratorPlugin now ([see here](https://github.com/aerogear/graphback/blob/master/packages/graphback-codegen-resolvers/src/ResolverGeneratorPlugin.ts#L83)) but it is only used to create the runtime services and is not used by the resolvers, so generation might be removed from here completely. This might make https://github.com/aerogear/graphback/issues/1028 obsolete.

@darahayes @wtrocki

**Comment by wtrocki** [comment]
Update. There is challenge of generating all services where only handful will be actually changed - making it maintenance burden on the user side. 
Also this change will affect only codegeneration where runtime extension will still be bad. We might start with spiking how to improve runtime service and db providers first and then see how we can adjust codegen to match it
